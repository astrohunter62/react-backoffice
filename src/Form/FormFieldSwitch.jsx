import React from 'react'
import PropTypes from 'prop-types'

import {
  FormControlLabel,
  FormHelperText,
  Switch,
} from '@material-ui/core'

class FormFieldSwitch extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    handleChange: PropTypes.func.isRequired,
    helperText: PropTypes.string,
    className: PropTypes.string,
    isDisabled: PropTypes.bool,
  }

  static defaultProps = {
    helperText: null,
    className: '',
    isDisabled: false,
  }

  state = {
    value: false,
  }

  handleChange(id) {
    const func = this.props.handleChange(id)

    return () => {
      const newValue = !this.state.value

      this.setState({
        value: newValue,
      })

      return func({
        target: {
          value: newValue,
        },
      })
    }
  }

  render() {
    const {
      id,
      title,
      helperText,
      isDisabled,
      className,
    } = this.props

    return (
      <div className={className}>
        <FormControlLabel
          disabled={isDisabled}
          control={
            <Switch
              checked={this.state.value}
              onChange={this.handleChange(id)}
              value={id}
              color="primary"
            />
          }
          label={title}
        />

        {helperText ? (
          <FormHelperText
            disabled={isDisabled}
          >
            {helperText}
          </FormHelperText>
        ) : null}
      </div>
    )
  }
}

export default FormFieldSwitch
