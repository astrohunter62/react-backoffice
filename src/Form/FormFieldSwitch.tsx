import React from "react";
import { FormControlLabel, FormHelperText, Switch } from "@material-ui/core";

type FormFieldSwitchProps = {
  id: string;
  title: string;
  handleChange: (...args: any[]) => any;
  helperText?: string;
  className?: string;
  isDisabled?: boolean;
};

type FormFieldSwitchState = {
  value: boolean;
};

class FormFieldSwitch extends React.Component<
  FormFieldSwitchProps,
  FormFieldSwitchState
> {
  state = {
    value: false,
  };

  handleChange(id: string) {
    const { handleChange } = this.props;
    const { value } = this.state;
    const func = handleChange(id);
    return () => {
      const newValue = !value;
      this.setState({
        value: newValue,
      });
      return func({
        target: {
          value: newValue,
        },
      });
    };
  }

  render() {
    const { id, title, helperText, isDisabled, className } = this.props;
    const { value } = this.state;
    return (
      <div className={className}>
        <FormControlLabel
          disabled={isDisabled}
          control={
            <Switch
              checked={value}
              onChange={this.handleChange(id)}
              value={id}
              color="primary"
            />
          }
          label={title}
        />

        {helperText ? (
          <FormHelperText disabled={isDisabled}>{helperText}</FormHelperText>
        ) : null}
      </div>
    );
  }
}

export default FormFieldSwitch;
