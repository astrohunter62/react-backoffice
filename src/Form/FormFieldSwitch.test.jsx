import React from 'react'
import renderer from 'react-test-renderer'

import FormFieldSwitch from './FormFieldSwitch'

it('renders correctly', () => {
  const tree = renderer
    .create((
      <FormFieldSwitch
        id="switch"
        title="Title"
        handleChange={() => () => {}}
      />
    ))
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders disabled', () => {
  const tree = renderer
    .create((
      <FormFieldSwitch
        id="switch"
        title="Title"
        handleChange={() => () => { }}
        isDisabled
      />
    ))
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders with helper text', () => {
  const tree = renderer
    .create((
      <FormFieldSwitch
        id="switch"
        title="Title"
        handleChange={() => () => { }}
        helperText="test"
      />
    ))
    .toJSON()
  expect(tree).toMatchSnapshot()
})
