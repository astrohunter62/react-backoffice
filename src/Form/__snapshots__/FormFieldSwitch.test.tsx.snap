// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <ForwardRef(WithStyles)
        checked={false}
        color="primary"
        onChange={[Function]}
        value="switch"
      />
    }
    label="Title"
  />
</div>
`;

exports[`renders disabled 1`] = `
<div>
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <ForwardRef(WithStyles)
        checked={false}
        color="primary"
        onChange={[Function]}
        value="switch"
      />
    }
    disabled={true}
    label="Title"
  />
</div>
`;

exports[`renders with helper text 1`] = `
<div>
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <ForwardRef(WithStyles)
        checked={false}
        color="primary"
        onChange={[Function]}
        value="switch"
      />
    }
    label="Title"
  />
  <WithStyles(ForwardRef(FormHelperText))>
    test
  </WithStyles(ForwardRef(FormHelperText))>
</div>
`;
